동기화 메모장 [STEP1] Jiseong, Toni, 애플사이다

yim2627 opened the pull request:
안녕하세요 콘! @protocorn93 

Jiseong, Toni, 애플사이다조 동기화 메모장 STEP1 완료하여 PR 제출드립니다!

iPad도 처음 만져보고, splitView도 처음 해봤는데.. 미숙하지만 잘 부탁드립니다! 😁

## STEP1 구현내용 및 고민한 점
### 1. 코드 구조
- Model
    - Memo 구조체 : JSON 데이터 parsing을 위해 맵핑할 Model 타입
- View
    - MasterTableViewCell 클래스 : Cell의 View 및 Layout과 관련된 기능을 구현, DateFormatter를 통해 지역화 구현
- Controller
    - MainSplitViewController 클래스 : `SplitView`를 커스텀함
    - MasterTableViewController 클래스 : `SplitView`의 `MasterView`. delegate를 가짐
    - DetailViewController 클래스 : `SplitView`의 `DetailView`. delegate 프로토콜을 채택
- 기타
    - SceneDelegate : Interface Buider 없이 코드로 UI를 구현하기 위해 스토리보드 삭제, `MainSplitViewController` 인스턴스 생성
     
### 2. 고민한 점
- MasterTableViewController의 delegate를 DetailViewController로 지정하기 위한 방법을 고민했습니다. Master/Detail View가 SplitView에 담겨있어 서로를 모르는 상황이므로 각 View에서 `프로퍼티 주입`을 하는 것보다는 아래 코드처럼 MainSplitViewController에서 `생성자 주입`을 하는 것이 낫다고 판단했습니다.
```swift 
private func configureUI() {
    //...
    detailViewController = DetailViewController()
    masterViewController = MasterTableViewController(style: .plain, delegate: detailViewController) // 생성자 주입

    self.viewControllers = [
        UINavigationController(rootViewController: masterViewController),
        UINavigationController(rootViewController: detailViewController)
    ]
}
```

- 상단 NavigationBarItem을 통해 primary view를 hide/show할 수 있게 하고, MainSplitViewController의 SplitBehavior/DisplayMode를 통해 가로/세로 모드를 변경해도 Column 형태가 유지되도록 했습니다.

## 궁금한 점

### 1. Cell Reuse 할 때 guard else문 처리방법
`dequeueReusableCell` 메서드를 사용할 때 cell을 custom Cell 클래스로 타입 캐스팅을 진행해주는데, 이 부분에서 옵셔널 바인딩을 항상 진행해왔었습니다.

위 과정에서 궁금했던 점은 해당 함수의 경우 Cell을 리턴하는데, guard else 처리시 그냥 빈 셀을 리턴해주는 것이 좋을지, 아니면 `fatalError`를 내보내는 것이 좋을지가 고민되었습니다.

저희는 `fatalError`로 설정하여 앱을 강제종료하는 것보단 차라리 빈 셀을 내보내는 것이 훨씬 안전하고 좋을 것 같다고 생각을 하였는데 콘의 생각은 어떠신지 궁금합니다.

아래 코드는 실제 셀을 dequeue하는 부분(`cellForRowAt`)에서 코드를 깔끔히 하고, 클래스명을 이용하여 identifier를 지정하려 생성해준 함수입니다.

```swift
extension UITableView {
    func dequeueReusableCell<T: UITableViewCell>(withClass name: T.Type, for indexPath: IndexPath) -> T {
        guard let cell = dequeueReusableCell(withIdentifier: String(describing: name), for: indexPath) as? T else {
            fatalError("Failed to dequeue cell.")  // 위험해보임
            // return T()  // 이것도 위험해보임
        }

        return cell
    }
}
```
        
### 2. contentOffset / setContentOffset 차이점
`TextView`에 Text를 넣은 후 프로그램을 실행 시켜보니, `TextView`의 Scroll이 최상단에 위치하지 않아, 나타난 글자들이 일부 잘리는 현상을 발견할 수 있었습니다. TextView의 `contentOffset`을 찍어보았는데 0,0임에도 해당 문제가 발견되었습니다. 겉보기에 해결이 되었고, 저희끼리 예상한 예상원인을 찾을 수 있었는데, font의 크기 때문이라고 생각을 했습니다. 코드의 작성순서를 `contentOffset`을 `.zero`로 설정한 후, font를 `LargeTitle`로 설정하면 text의 글자가 잘리는 모습을 볼 수 있었고, 순서를 반대로 작성하면 정상적으로 보이는 것을 발견할 수 있었습니다. 이에 저희가 내렸던 결론은 `contentOffset`이 먼저 0,0이 된 후, 글자가 커졌기에 발생하는 현상이라고 결론을 내렸습니다. 

저희가 문제를 해결하는 도중, `setContentOffset`과 `contentOffset`이 혼동되어 질문드립니다. 다음은 저희가 실험을 했던 코드들입니다.

```swift
textView.contentOffset = .zero  // 문제해결

textView.setContentOffset(.zero, animated: false)  // 적용 안됨
textView.setContentOffset(CGPoint(x: 0, y: 1, animated: false)  // 스크롤이 최상단으로 이동
textView.setContentOffset(CGPoint(x: 0, y: 100, animated: false)  // 100만큼 offset되지 않음
```
특히 y가 1이면 스크롤이 최상단으로 이동하는데, y가 100일 때 100만큼 이동하지 않아서 왜 이런 현상이 발생하는지 궁금합니다.


그래서 공식문서를 찾아보았는데, 다음의 내용을 볼 수 있었습니다.

- contentOffset : ScrollView의 origin에 대한 ContentView의 origin
- setContentOffset : Sets the offset from the content view’s origin that corresponds to the receiver’s origin. ← 여기에서 ViewControll의 View `receiver` (addSubview를 호출하는 부모 뷰)가 무엇인지도 잘 모르겠습니다.

### 3. parsing한 데이터를 Detail View에 전달하는 방법

JSON 파일을 파싱한 데이터를 tableView에서도 사용을 하고, tableView의 특정 셀을 눌렀을 때 나오는 DetailView에서도 사용을 해야하는데, 파싱을 한번만 하여 나온 데이터를 두 곳 모두에서 사용해주려면 어떤 방법을 사용해야 가장 효율적일지가 궁금합니다.

저희가 생각한 방법은 아래와 같습니다.
- 방법-1. SplitView의 showDetailViewController 메서드 활용 <- 선택
- 방법-2. ViewController의 showDetailViewController 메서드 활용 
- 방법-3. ViewController Factory를 생성하여 열거형의 연관값을 활용

처음 생각한 방법은 Factory을 생성하는 것이었는데, 이는 `DetailViewController`를 생성하는 Factory이기 때문에 특정 셀이 클릭될 때마다 뷰컨트롤러가 새로이 생성되는 큰 문제점이 있을 것 같아 진행하지 않았습니다.

그래서 결과적으로 방법-1의 `showDetailViewController` 메서드를 활용하여 하나의 뷰컨트롤러의 인스턴스를 미리 생성한 뒤 뷰에 올려진 텍스트뷰의 텍스트를 업데이트 해주는 방법으로 구현하였는데, 저희가 사용한 방식이 괜찮은 방식인지, 그리고 현업에서 이러한 상황일 시 어떤 방식을 주로 활용하는지가 궁금합니다.

### 4. Cell의 날짜 Label
앱 사용 중에 사용자가 시스템의 지역을 바꾸면, 한국은 `연/월/일`, 미국은 `월/일/연` 순으로 표시하도록 하기 위해 dateFormatter의 `dateFormat`을 사용하지 않고, `dateStyle` 을 사용했습니다.

(`dateFormat`을 사용하지 않은 이유는 형식을 고정시켜버려 시스템 지역을 변경하여도 나라에 따른 날짜 표기 방식이 변하지 않는 문제점이 있었습니다.)

그런데 `dateFormat`을 사용하지 않으니 프로젝트 요구사항 예시처럼 `.`이 나오지않고 `/`만 나오는 문제점이 있었는데, dot을 replacingOccurence 메서드를 활용하여 `/`으로 대체하는 방법 밖에 없는지 궁금합니다.

### 5. delegate 패턴
tableViewController가 가진 데이터를 DetailViewController에게 전달할 때 tableViewController가 DetailViewController를 알아야했는데, 이때 결합도를 낮추기 위해 delegate 패턴을 사용하였습니다. 

하지만 `showDetailViewController` 메서드를 사용하는 과정에서 `DetailViewController`를 무조건 알아야 show를 할 수 있을 것 같아 이렇게 작성하였는데(6번째 줄), 타입 캐스팅을 해줌으로써 테이블뷰컨트롤러가 다시 `DetailViewController`을 직접적으로 알게되는 것인지, 정상적으로 코드를 작성한 것인지 궁금합니다.

```swift
override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        guard let memos = memos else {
            return
    }
        
    guard let destination = delegate as? DetailViewController else { // <-
        return
    }
    
    destination.applyData(with: memos[indexPath.row].description)
    splitViewController?.showDetailViewController(UINavigationController(rootViewController: destination), sender: self)
}
```

## 조언을 듣고싶은 점

전체적인 가독성이나 코드 구조가 괜찮은지 궁금합니다!

oksunwoo committed:
test: pod test

just1103 committed:
test: 코드 추가

just1103 committed:
docs: git ignore 파일에 Pods 폴더 추가

yim2627 committed:
feat: swiftlint 컨벤션 추가 #1

just1103 committed:
feat: MainSplitViewController 추가 #1

- 사용자 정의 init 추가
- Storyboard 삭제 및 SceneDelegate의 rootViewController 변경

oksunwoo committed:
feat: MasterTableView 구현 #1

- extension으로 dequeueReusableCell메서드 구현
- CustomCell 구현

just1103 committed:
feat: DetailViewController 추가 #1

- DetailViewController의 TextView 구현
- SplitView에 Master/Detail View 연결
- 폴더 구분

yim2627 committed:
feat: NavigationBar 구현 #1

- Navigation Title, Button 구현
- Dynamic Type 적용

oksunwoo committed:
feat: JSON Data 파싱 #1

- Memo 모델 생성
- fetchData() 메서드를 tableView에 추가

just1103 committed:
feat: MasterTableViewCell UI Layout 구현 #1

- configureUI, applyData 메서드 추가
- 접근제어자 추가

yim2627 committed:
feat: DetailViewController textView 업데이트되도록 구현 #1

- Delegate 패턴 적용
- MemoSelectionDelegate 프로토콜 구현

oksunwoo committed:
feat: dateFormatter 구현 #1

- 컨벤션 수정
- 접근제어자 수정

protocorn93 commented on CloudNotes/CloudNotes/Controller/DetailViewController.swift:
```diff
@@ -0,0 +1,42 @@
+        textView.font = .preferredFont(forTextStyle: .body)
+        textView.contentOffset = .zero
+        
+        textView.translatesAutoresizingMaskIntoConstraints = false
```
`MasterTableViewCell` 처럼 `textView` 정의부에서 이러한 속성들을 지정해주고 `configureTextView`에서는 레이아웃 코드만 위치하는게 전체적으로 보다 더 좋은 가독성과 일관성을 가질 수 있을 것 같습니다.

just1103 replied:
textView도 정의부에서 속성을 지정하도록 수정했습니다. 통일성, 가독성이 개선된 것 같아요. 👍

protocorn93 commented on CloudNotes/CloudNotes/Controller/MainSplitViewController.swift:
```diff
@@ -0,0 +1,32 @@
+        preferredDisplayMode = .oneBesideSecondary
+        
+        detailViewController = DetailViewController()
+        masterViewController = MasterTableViewController(style: .plain, delegate: detailViewController)
```
이 두 뷰컨트롤러 객체 역시 외부에서 의존성으로 생성자를 통해 전달받도록 하는건 어떨까요? 변수도 `var`가 아닌 `let`으로, 논옵셔널 타입으로 가져갈 수 있으며 각각의 뷰컨트롤러 생성 로직이 `MainSplitViewController`안에 강하게 결합되는 것도 방지할 수 있을 것 같아요.

just1103 replied:
말씀하신 것처럼 SceneDelegate에서 ChildViewController를 생성하고, MainSplitViewController의 생성자 주입으로 전달하도록 개선했습니다. 이러한 개선방향이 맞을까요?
그리고 결합도를 낮추면 자식 View를 변경하는 등 확장성 측면에서 유리한 것 같아보이는데, 다른 이점도 있는지 궁금합니다!

protocorn93 replied:
아래에도 비슷하게 언급을 했지만, 현재는 뷰컨트롤러 안에 모든 코드들이 담겨있어요. 하지만 코드를 각각의 역할별로 분리해볼 수 있습니다. 이렇게 역할별로 분리가 되고, 이런것들이 정형화, 패턴화된 것들이 일종의 아키텍처입니다. (MVVM, VIP, RIBs 등등)

이런 아키텍처들의 특징은 느슨한 결합과 대체 가능성으로 확장성 있고 유연한 설계를 할 수 있다는 장점이 있어요. 아직은 와닿지 않으실 수 있으시겠지만 객체와 객체간의 강결합을 지양하는 연습을 하신다고 생각하면 좋을 것 같아요!

protocorn93 commented on CloudNotes/CloudNotes/Controller/MasterTableViewController.swift:
```diff
@@ -0,0 +1,87 @@
+    private var memos: [Memo]?
+    weak var delegate: MemoSelectionDelegate?
+    
+    // MARK: - Methods
```
사소하지만 `Methods`와 `Life cycle`은 구분해서 마크업해주는게 좋지 않을까 싶어요!

oksunwoo replied:
네 `LifeCycle`도 마크업 해주었습니다.

protocorn93 commented on CloudNotes/CloudNotes/Controller/MasterTableViewController.swift:
```diff
@@ -0,0 +1,87 @@
+        
+        guard let data = try? String(contentsOfFile: file).data(using: .utf8) else {
+            return
+        }
```
`Bundle.main.path`를 사용해서 경로를 얻어와 `String`으로 변환하고 이를 다시 `Data` 타입으로 변환 시키는 과정으로 데이터를 추출하고 있는데, 

`Bundle.main.url`로 `URL`을 얻어와 `Data(contentsOf:)` 생성자로 바로 `Data` 타입을 만들어줄 수도 있는데 위와 같은 선택을 한 이유가 궁금합니다.

oksunwoo replied:
위와 같은 선택을 한 이유는 그러한 방법이 있는지 몰랐습니다 😂
그래서 다음과 같이 변경해주었습니다 🙂
```swift
guard let url = Bundle.main.url(forResource: "sample", withExtension: "json") else {
            return nil
        }
        
        guard let data = try? Data(contentsOf: url) else {
            return nil
        }
```

protocorn93 commented on CloudNotes/CloudNotes/Controller/MasterTableViewController.swift:
```diff
@@ -0,0 +1,87 @@
+
+class MasterTableViewController: UITableViewController {
+    // MARK: - Properties
+    private var memos: [Memo]?
```
빈 배열이 아닌 옵셔널로 정의한 이유가 궁금합니다.

just1103 replied:
"nil 가능성"이 있음을 내포하고 있어서 옵셔널로 정의했습니다.
해당 변수에 JSONDecoder를 통해 decodind한 데이터를 할당하도록 했는데, decoding 과정에서 에러가 발생하면 try?를 통해 nil을 반환하도록 하기 때문입니다. 
이 경우 옵셔널로 둬도 괜찮을까요?

protocorn93 replied:
네~ 경우에 따라서 `nil`과 `[]`을 구분해야할 필요가 있죠~ 하지만 저는 보통 이런 구분은 비즈니스 로직을 담당하는 객체에서 구분합니다.

`MasterTableViewController`에서 화면에 표현되는 값을 나타내는 `Memo` 같은 경우에는 `[]`로 두는 것이 더욱 명확하지 않나 싶어요. 아직은 UI 로직과 비즈니스 로직을 분리하는데 익숙치 않으실수있습니다. 이는 추후 프로젝트를 진행하면서 이 둘을 분리해보는 연습을 통해 익혀나가시면 될 것 같습니다.

protocorn93 commented on CloudNotes/CloudNotes/Extensions/UITableView+ReuseCell.swift:
```diff
@@ -0,0 +1,11 @@
+import UIKit
+
+extension UITableView {
+    func dequeueReusableCell<T: UITableViewCell>(withClass name: T.Type, for indexPath: IndexPath) -> T {
+        guard let cell = dequeueReusableCell(withIdentifier: String(describing: name), for: indexPath) as? T else {
+            return T()
+        }
+        
+        return cell
```
`extensions`을 활용해 보다 쉽게 셀을 dequeue해주도록 한 점 훌륭합니다!

protocorn93 commented on CloudNotes/CloudNotes/Controller/MasterTableViewController.swift:
```diff
@@ -0,0 +1,87 @@
+    }
+    
+    private func configureTableView() {
+        tableView.register(MasterTableViewCell.self, forCellReuseIdentifier: String(describing: MasterTableViewCell.self))
```
`String(describing: MasterTableViewCell.self)` 이 코드도 `UITableViewCell`의 `extension`으로 빼볼 수 있지 않을까요? 

`MasterTableViewCell.reuseIdentifier`처럼 좀 더 쉽게 사용해볼 수 있을 것 같아요!

just1103 replied:
말씀하신 것처럼 UITableViewCell 익스텐션으로 `reuseIdentifier`를 static 연산 프로퍼티로 추가했습니다.
사용부의 가독성이 훨씬 좋아진 것 같아요.

protocorn93 commented on CloudNotes/CloudNotes/View/MasterTableViewCell.swift:
```diff
@@ -0,0 +1,89 @@
+
+class MasterTableViewCell: UITableViewCell {
+    // MARK: - Properties
+    private let entireStackView: UIStackView = {
+        let stackView = UIStackView()
+        stackView.axis = .vertical
+        stackView.distribution = .fill
+        stackView.alignment = .fill
+        stackView.spacing = 10
+        stackView.translatesAutoresizingMaskIntoConstraints = false
+        
+        return stackView
+    }()
```
코드로 뷰를 정의하는 블록에서 뷰의 속성도 모두 지정해주고 반환해주니 레이아웃을 하는 코드가 더욱 깔끔해질 수 있었던 것 같아요! 👍

protocorn93 commented on CloudNotes/CloudNotes/Controller/DetailViewController.swift:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    func configureUI() {
+        view.backgroundColor = .white
```
저는 개인적으로 메서드안의 추상화 레벨을 맞추는걸 선호합니다. 그래서 한 줄 짜리 코드라도 `configureContentView` 혹은 `configureRootView`와 같은 네이밍으로 뷰컨트롤러의 루트 뷰 속성을 정해주는 메서드도 따로 분리하곤합니다.

just1103 replied:
configureContentView 메서드를 추가하여 개선했습니다. 👍

protocorn93 reviewed (changes requested):
세 분 모두 수고하셨습니다. 전체적으로 깔끔한 가독성을 갖고 있는 코드로 읽기에 너무 편안했어요! 전반적으로도 크게 모나지 않은 코드로 보여집니다! 

## 구현내용 및 고민한 점 

### 코드 구조 

저도 작은 프로젝트를 진행할 때는 여러분이 작성한 디렉토리 구조를 따릅니다. 하지만 조금만 앱이 커지고 기능이 많아지게 되면 이러한 구조는 원하는 파일을 쉽게 찾기 힘들다는 점도 알고 계시면 좋을 것 같아요. 저는 그래서 보통 기능이 많아지거나 화면이 많아지면 화면 단위로 관련 파일들을 묶어두곤해요. 

PS. 그리고 `sample.json`은 `LaunchScreen.storyboard`, `assets`와 함께 `Resources`라는 디렉토리로 묶어두곤합니다. 

---

## 궁금한 점 

### 1. Cell Reuse 할 때 guard else문 처리방법

저는 최대한 강제 언랩핑은 지양하는 편이에요. 그렇기 때문에 세 분이 사용한 방법처럼 빈 셀을 반환해주곤 합니다. 다만 최근엔 DiffableDataSource를 사용하고 있고 이를 사용하면 `nil`을 반환할 수 있는 코드 구조이기에 빈 셀을 반환하는 코드는 작성하지 않을 수 있었어요.

### 2. contentOffset / setContentOffset 차이점

먼저 제가 실행한 코드에선 `contentOffset`에 별도의 값을 지정하지 않아도 텍스트는 잘리지 않습니다. 혹시 재현된 화면을 캡처해주실 수 있으신가요?? 환경도 같이 알려주세요. 저는 iPad Pro 9.7인치 시뮬레이터에 테스트해보았습니다.`largeTitle`을 지정해도 잘리지 않네요.

그리고 `configureTextView`가 호출되는 부분은 `viewDidLoad`입니다. 뷰컨트롤러의 라이프사이클을 고려해서 적절한 메서드에서 contentOffset을 조정해보세요.  또한 `contentOffset`에 값을 지정하는 것과 `setContentOffset` 메서드를 통해 값을 지정하는 것은 다르지 않습니다. 애니메이션을 가능케하냐 못하냐의 차이입니다. 

`setContentOffset`의 정의에서 `receiver`가 의미하는바는 수신자, 즉 `setContentOffset`이라는 메시지를 받는 수신자를 의미합니다.  왜 이를 메시지라 표현하고 수신자라고 표현하는지는 객체지향과 다형성에 대한 이해가 필요해요. 

단순히 "메서드를 호출했다"와 메시지를 전달해서 이 메시지에 응답했다는 엄연히 다릅니다. 왜냐하면 메서드 호출은 특정 기능을 구현한 메서드를 콕 집어서 호출한 것이고, 메시지를 전달하고 이에 응답했다는 것은 어느 객체가 응답할지 모르는 상황이기 때문입니다. 

`UIScrollView`를 상속받은 다양한 클래스들의 `setContentOffset`를 오버라이드 할 수 있습니다. 그리고 그 동작은 달라질 수 있습니다. 그렇기 때문에 `setContentOffset`이라는 메시지를 받아 각자만의 방식으로 이 메시지에 대해 응답(오버라이드, 구현)하고 있는겁니다. 

그렇기 때문에 `setContentOffset` 메서드 정의의 `receiver's`는 `UITextView` 인스턴스가 될수도 있고, `UITableView`, `UICollectionView` 인스턴스가 될 수 있습니다. 현재 프로젝트에선 `UITextView` 인스턴스죠


### 3. parsing한 데이터를 Detail View에 전달하는 방법
현재 스탭에 집중한다면 현재의 방법도 무난해보입니다. 다만 메모앱이라면 추후 편집이 들어갈텐데, 이때 상세 화면을 하나의 뷰컨트롤러로 사용한다면 데이터를 수정하고 저장하고 관리하는데 더 많은 공수가 들어가지 않을까라는 생각이 듭니다. 

그렇기 때문에 셀을 클릭했을 때 상세 뷰컨트롤러를 생성해주고 이때 둘의 관계를 구성해주고 띄우는 방식이라면 그 둘은 하나의 데이터로 묶여있는 관계이기 때문에 관리가 더욱 용이하지 않을까 하는 생각이듭니다. 

### 4. Cell의 날짜 Label
`DateFormatter`의 `setLocalizedDateFormatFromTemplate` 메서드를 사용해보면 어떨까요? 

`setLocalizedDateFormatFromTemplate("yyyy.MM.dd")` 이렇게 지정했을 때
한국이라면 2021.02.10
미국이라면 10/02/2021
이렇게 노출됩니다. 

`setLocalizedDateFormatFromTemplate("dd/MM/yyyy")` 이렇게 지정했을 때 
한국이라면 2021.02.10
미국이라면 10/02/2021
이렇게 노출됩니다. 

이를 통해 추측해본 결과로는 한국의 시간 표현의 표준에선 `.`으로 구분하고 미국의 시간 표현의 표준에선 `/`으로 구분하고 있는 것으로 보입니다. 
### 5. delegate 패턴

몇 가지 방법이 떠오르네요. 

1. `MemoSelectionDelegate` 프로토콜 안에 `var memoSelectionDestination: UIViewController { get }` 프로퍼티를 정의해볼 수 있을 것 같아요. 그러면 채택을 한 뷰컨트롤러에서 `return self` 해주면 되지 않을까요? 

2. `weak var delegate: (MemoSelectionDelegate & UIViewController)?` 타입으로 프로퍼티를 정의할 수 있을 것 같아요. 다만 1번의 방법보다 현재 컨텍스트에 보다 강하게 결합되어 재사용성이 떨어질 수 있을 것 같긴합니다.

yim2627 committed:
style: 폴더 구조 수정 #2

oksunwoo committed:
refactor: configureTextView의 코드 분리 #2

just1103 committed:
refactor: MainSplitViewController의 생성자 주입을 통한 ChildView 지정 #2

- SceneDelegate에서 ChildViewController를 생성
- MainSplitViewController 초기화 시 ChildViewContoller를 지정

yim2627 committed:
refactor: masterTableViewController DataSource 분리 #2

- 상속하지 않는 클래스의 경우 final 키워드 추가

oksunwoo committed:
refactor: Bundle에서 Data를 받아오는 방식 변경 #2

- Mark 주석 추가
- configureTextView 코드 분리

just1103 committed:
refactor: UITableViewCell 익스텐션으로 reuseIdentifier 프로퍼티 추가 #2

- DetailViewController의 configureUI 메서드 추가를 통해 추상화 레벨 통일

yim2627 commented:
피드백 주신 내용을 반영해보며, 궁금한 점이 생겨 다시 질문드립니다.

### 3. parsing한 데이터를 Detail View에 전달하는 방법
    
> 메모앱이라면 추후 편집이 들어갈텐데, 이때 상세 화면을 하나의 뷰컨트롤러로 사용한다면 데이터를 수정하고 저장하고 관리하는데 더 많은 공수가 들어가지 않을까라는 생각이 듭니다. 그렇기 때문에 셀을 클릭했을 때 상세 뷰컨트롤러를 생성해주고 이때 둘의 관계를 구성해주고 띄우는 방식이라면 그 둘은 하나의 데이터로 묶여있는 관계이기 때문에 관리가 더욱 용이하지 않을까 하는 생각이듭니다.
    
라고 코멘트 주신 부분에 대해 고민해봤는데, 이해가 어려운 점이 있었습니다.
    
1. STEP2를 고려하더라도 DetailViewController의 Subview가 titleLabel, textView 2개 뿐이라서 데이터를 관리 (수정, 저장, 삭제 등)할 때 공수가 더 많이 들어간다는 점이 잘 이해가 되지 않습니다.
        
저희가 생각한 것은 하나의 뷰컨트롤러 인스턴스의 TextView의 Text를 업데이트 해주는 방식이 뷰컨트롤러 인스턴스를 특정 셀이 눌릴 때마다 생성하고 지우는 것보다 더 효율적일 것이라고 생각하여 코드를 작성했습니다.
        
2. 선택한 Cell과 DetailViewController를 하나로 묶으면 새로운 셀이 생성될 때마다 DetailViewController 인스턴스가 계속 증가하기 때문에 메모리에 부담이 되지 않을까 우려되는 점도 있었는데, 어떤 점에서 관리가 용이할지 궁금합니다.

### 4. Cell의 날짜 Label
    
`setLocalizedDateFormatFromTemplate`을 찾아봤는데 반드시 locale을 지정한 이후에 사용하라고 명시되어 있었습니다. 이때 기존의 `.autoupdatingCurrent`은 locale에 반영되지 않는 것으로 파악했습니다. 특정 국가의 locale을 할당해야 반영이 되던데, 콘도 국가별 locale을 지정하셨던 건지 궁금합니다.

```swift
private let dateFormatter: DateFormatter = {
    let dateFormatter = DateFormatter()
//    dateFormatter.dateStyle = .short
    
    dateFormatter.timeZone = .autoupdatingCurrent 
//    dateFormatter.locale = .autoupdatingCurrent // 기존 코드 - 반영되지 않는 문제 발생
    
    dateFormatter.locale = Locale(identifier: "ko_KR")   // 특정 국가의 locale을 넣어줘야 반영됨 (2020.12.23)
//    dateFormatter.locale = Locale(identifier: "en_US") // (12/23/2020)
    
    dateFormatter.setLocalizedDateFormatFromTemplate("yyyy.MM.dd") // 자동으로 업데이트할 수 없는 메서드라서 autoupdatingCurrent를 반영 못하는듯 합니다
    
    return dateFormatter
}()
```

### 5. delegate 패턴

방법-1. didSelectRowAt 내부에서 타입캐스팅, 그리고 방법-2. 프로토콜 내부에서 DetailVC 타입 명시, 이 중에서 어떤게 결합도가 낮을지 궁금합니다.

```swift
// 방법-1. didSelectRowAt 내부에서 타입 캐스팅
protocol MemoSelectionDelegate: AnyObject {
    var memoselectionDestination: UIViewController { get }
}

guard let destination = delegate?.memoselectionDestination as? DetailViewController else {
    return
}

// 방법-2. 프로토콜 내부에서 DetailVC 타입 명시
protocol MemoSelectionDelegate: AnyObject {
    var memoselectionDestination: DetailViewController { get }
}

guard let destination = delegate?.memoselectionDestination else {
    return
}
```
    
- 추가한 내용) 역할 분리를 위해 TableView의 TableViewDataSource를 별도 타입으로 분리하고, JSON fetch 데이터도 DataSource에서 관리하도록 개선했습니다. 적절한지 궁금합니다.

yim2627 committed:
refactor: MemoSelectionDelegate에 UIViewController 타입의 프로퍼티 추가 #2

oksunwoo committed:
feat: Keyboard가 텍스트뷰를 가리지 않도록 구현 #1

- Notification 추가

yim2627 committed:
refactor: tableView style 변경 및 selectedBackgroundColor 설정 #2

yim2627 committed:
Merge branch 'step1' of github.com:just1103/ios-cloud-notes into step1

protocorn93 commented:
## 3. parsing한 데이터를 Detail View에 전달하는 방법

네 이 부분은 제가 복잡하게 생각했던 것 같습니다. 지금 현재 방식 그대로 유지해보시는게 좋아보입니다. 

다만 2번, 셀을 클릭할 때마다 인스턴스가 쌓일 것 같지는 않은데, 이 부분도 실험해볼만한 주제인 것 같네요! 

## 4. Cell의 날짜 Label

아하 이 부분은 제가 놓친 것 같네요. 음.. 이건 저도 좀 실험을 해봐야할 것 같습니다. 일단은 현행유지해주시면 될 것 같아요

## 5. delegate 패턴

네 제가 생각했을 땐 `UIViewController`가 보다 유연한 방법인 것 같아요. 이를 상속받은 다른 뷰컨트롤러를 사용한다 가정했을 때 (예를 들어 글의 형식에 따라 UI가 많이 상이할 경우 하나의 뷰컨트롤러에서 타입에 따른 분기를 치기보다는 별도의 다른 뷰컨트롤러를 사용해볼 수 있지 않을까하는 생각도드네요.) 보다 유연하게 대처해볼 수 있을 것 같아요.

> 추가한 내용) 역할 분리를 위해 TableView의 TableViewDataSource를 별도 타입으로 분리하고, JSON fetch 데이터도 DataSource에서 관리하도록 개선했습니다. 적절한지 궁금합니다.

네!! 이렇게 역할을 분리해내는 것부터 시작하는 겁니다! 매우 좋은 시도인 것 같아요!!

protocorn93 commented on CloudNotes/CloudNotes/Controller/MasterTableViewController.swift:
```diff
@@ -0,0 +1,61 @@
+
+final class MasterTableViewController: UITableViewController {
+    // MARK: - Properties
+    let memoDataSource = MasterTableViewDataSource()
```
`MasterTableViewDataSource`를 프로토콜로 추상화하고, 이를 `MasterTableViewController` 생성시점에 주입해주는게 어떨까요? 그러면 mock 객체를 넣어 테스트도 용이할 것 같네요! 

```swift
protocol MasterTableViewDataSource: UITableViewDataSource { 
  var memos: [Memo]? { get }
}
```

just1103 replied:
MasterTableViewDataSourceProtocol을 추가하고, MasterTableViewController의 생성자 주입을 통해 DataSource 객체를 할당했습니다. 추상화하니까 결합도를 또 개선할 수 있네요!

oksunwoo committed:
refactor: MasterTableViewDataSource를 프로토콜로 추상화 #2
